### PR TITLE
[CodeCompletion] Improve context type analysis for dictionary literals

### PR DESCRIPTION
**Pull request by rintaro**
- Analyze the type of the literal in the context
- If `:` is missing in the literal, treat the expression as a key expression
- If the parent expression is a `TupleExpr`, analyze the context type of the tuple first, then return the element type of the position
- Use `CodeCompletionExpr` as a value expression in the dictionary literal if the value is missing. This increases the chance to type check them successfully.

rdar://problem/57096392

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/ExprContextAnalysis.cpp [hunk 1]
```diff
@@ -602,9 +602,46 @@ class ExprContextAnalyzer {
...
+          // let _: [Key : Value] = [#HERE#]
+          // In this case, 'Key' is the expected type.
+          if (boundGenericT->getDecl() == Context.getDictionaryDecl())
+            recordPossibleType(boundGenericT->getGenericArgs()[0]);
+        }
+      }
+      break;
+    }
+    case ExprKind::Dictionary: {
+      // Check context types of the array literal expression.
```
Comment (benlangmuir) [inline lib/IDE/ExprContextAnalysis.cpp]: array -> dictionary?

**Review by benlangmuir — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/ExprContextAnalysis.cpp (hunk 1 above)
Comment (rintaro) [reply lib/IDE/ExprContextAnalysis.cpp]: 🤦‍♂ Thanks!

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**